kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -146,28 +115,59 @@ else if (requestFilterResult.message() != null) {
         }
     }
 
+    private CompletableFuture<RequestFilterResult> executeWrite(ChannelHandlerContext ctx, DecodedRequestFrame<?> decodedFrame,
+                                                                DefaultFilterContext filterContext,
+                                                                CompletableFuture<RequestFilterResult> withTimeout) {
+        return withTimeout.whenComplete((requestFilterResult, t) -> {
+            // maybe better to run the whole thing on the netty thread.
+
+            if (t != null) {
+                LOGGER.warn("{}: Filter{} for {} request ended exceptionally - closing connection",
+                        ctx.channel(), filterDescriptor(), decodedFrame.apiKey(), t);
+                filterContext.closeConnection();
+                return;
+            }
+
+            if (requestFilterResult.drop()) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("{}: Filter{} drops {} request",
+                            ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                }
+            }
+            else if (requestFilterResult.message() != null) {
+                if (requestFilterResult.shortCircuitResponse()) {
+                    // this is the short circuit path
+                    var header = requestFilterResult.header() == null ? new ResponseHeaderData() : ((ResponseHeaderData) requestFilterResult.header());
+                    header.setCorrelationId(decodedFrame.correlationId());
+                    filterContext.forwardResponseInternal(header, requestFilterResult.message());
+                }
+                else {
+                    var header = requestFilterResult.header() == null ? decodedFrame.header() : requestFilterResult.header();
+                    filterContext.forwardRequestInternal((RequestHeaderData) header, requestFilterResult.message());
+                }
+            }
```

Comment:
Appreciate this is probably just a code shuffle, but are we missing an `else` clause here?

Reply:
It is a consequence of the valid combinations of drop/close/forward.  I think the code could be expressed slightly better. Let me put up a PR.

Reply:
https://github.com/robobario/kproxy/pull/8